### DISCUSSION TITLE
Class template argument deduction for `absl::AnyInvocable`?

### DISCUSSION THREAD

**Discussion by jiawen** [abseil/abseil-cpp#1755] [General]
C++17 added CTAD to std::function by way of a [deduction guide](https://eel.is/c++draft/func.wrap.func.con#itemdecl:6). This permits writing:

`std::function f = []() { return 3; };`

But `absl::AnyInvocable` does not current have this facility. E.g., the following fails to compile:

`absl::AnyInvocable g = []() { return 3; };`

Is there thought to adding such a facility?

**Comment by higher-performance** [comment]
The closer comparison to make here is with `std::move_only_function`, not `std::function`. It does _not_ have deduction guides: https://en.cppreference.com/w/cpp/utility/functional/move_only_function

This is likely intentional, not an oversight -- both `std::move_only_function` and `absl::AnyInvocable` support distinguishing `void()` from `void()&&`, for example, but those can't be inferred.

I would also argue CTAD here hurts readability.
